### PR TITLE
[Feature] Update the calling hero to use the Azure Calling UI Library

### PR DESCRIPTION
**Pull request by Mohtasim**
## Purpose

* Hero sample migration using communication ui calling library
* Most of the UI is supported with Fluent UI Components
* AAD is configured with the sample 
* The sample further uses Graph API to fetch user account information
* Error handling events are shown on the UI

## Does this introduce a breaking change?

```
[x] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[x] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Clone the code
*  Setup the environment and add configuration as per the ReadMe.md

## What to Check
Verify that the following are valid
* Group call is started
* Able to join a group meeting or teams meeting
* Check whether the error surfaces properly

### CODE REVIEW

**Review by ShaunaSong — APPROVED** [review]

File: AzureCalling/app/src/androidTest/java/com/azure/samples/communication/ui/calling/ExampleInstrumentedTest.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+@RunWith(AndroidJUnit4.class)
+public class ExampleInstrumentedTest {
+    @Test
+    public void useAppContext() {
+        // Context of the app under test.
+        final Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
+        assertEquals("com.azure.samples.communication.ui.calling", appContext.getPackageName());
+    }
+}
```
Comment (iaulakh) [inline AzureCalling/app/src/androidTest/java/com/azure/samples/communication/ui/calling/ExampleInstrumentedTest.java:31]: We can remove test classes and gradle dependencies as in previou sample

File: AzureCalling/app/build.gradle [hunk 2]
```diff
@@ -61,13 +59,15 @@ android {
...
+    implementation 'androidx.appcompat:appcompat:1.4.2'
+    implementation 'com.google.android.material:material:1.6.1'
+    implementation 'androidx.constraintlayout:constraintlayout:2.1.4'
+    implementation 'com.azure.android:azure-communication-ui-calling:+'
+    implementation 'com.microsoft.identity.client:msal:+'
+    implementation 'com.microsoft.fluentui:FluentUIAndroid:+'
+    implementation 'com.android.volley:volley:1.2.1'
+    testImplementation 'junit:junit:4.13.2'
+    androidTestImplementation 'androidx.test.ext:junit:1.1.3'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
```
Comment (iaulakh) [inline AzureCalling/app/build.gradle:72]: We can remove test dependencies

**Review by iaulakh — CHANGES_REQUESTED** [review]
Lanscape mode can not join a call

Screen Shot 2022-08-26 at 10 54 59 AM

Landscape mode sign in button is covered
Screen Shot 2022-08-26 at 10 55 18 AM

Display name is null
Screen Shot 2022-08-26 at 10 55 50 AM

In dark mode Name text line not visible
Screen Shot 2022-08-26 at 10 57 07 AM

Portrait mode
Screen Shot 2022-08-26 at 10 58 31 AM

Group meeting not visible in dark mode
Screen Shot 2022-08-26 at 10 59 38 AM

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/contracts/SampleErrorMessages.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+public class SampleErrorMessages {
+    public static final String GROUP_ID_REQUIRED = "Group call ID is required";
+    public static final String GROUP_ID_INVALID = "Group call ID is not valid";
+    public static final String DISPLAY_NAME_REQUIRED = "Display Name is required";
+    public static final String TEAMS_LINK_REQUIRED = "Teams Link is required";
+    public static final String TEAMS_LINK_INVALID = "Teams Link is not valid";
+    public static final String CALL_COMPOSITE_JOIN_CALL_FAILED = "Could not join call";
+    public static final String CALL_COMPOSITE_END_CALL_FAILED = "Could not end call";
+    public static final String CALL_COMPOSITE_TOKEN_EXPIRED = "ACS Token has expired";
+}
```
Comment (iaulakh) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/contracts/SampleErrorMessages.java:15]: Please create task to refactor to use XML.

File: AzureCalling/app/src/main/res/layout/fragment_group_meeting.xml [hunk 4]
```diff
@@ -0,0 +1,109 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Mrayyan) [inline AzureCalling/app/src/main/res/layout/fragment_group_meeting.xml:1]: Missing copyright tag

File: AzureCalling/app/src/main/res/layout/fragment_teams_meeting.xml [hunk 5]
```diff
@@ -0,0 +1,125 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (Mrayyan) [inline AzureCalling/app/src/main/res/layout/fragment_teams_meeting.xml:1]: Missing copyright tag. You can add to all the other xml files as well that are missing it.

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/externals/authentication/AADAuthHandler.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+            }
+
+            @Override
+            public void onError(final MsalException exception) {
+                Log.e(LOG_TAG, exception.getMessage());
+            }
+
+            @Override
+            public void onCancel() {
+                Log.d(LOG_TAG, "User cancelled login.");
```
Comment (iaulakh) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/externals/authentication/AADAuthHandler.java:80]: should be from XML

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/externals/authentication/AADAuthHandler.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+
+    private void acquireToken(final Activity activity, final Consumer<Object> callback) {
+
+        mSingleAccountApp.acquireToken(activity, mScopes, new AuthenticationCallback() {
+            @RequiresApi(api = Build.VERSION_CODES.N)
+            @Override
+            public void onSuccess(final IAuthenticationResult authenticationResult) {
+                accessToken = authenticationResult.getAccessToken();
+                editor.putString(ACS_ACCESS_TOKEN, accessToken);
+                editor.apply();
```
Comment (iaulakh) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/externals/authentication/AADAuthHandler.java:161]: Inject as object. AAD should not be saved into shared preferences.

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/externals/authentication/AADAuthHandler.java [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+
+        if (accessToken == null || accessToken.length() == 0) {
+            findUserProfile(activity, authCallback);
+        }  else {
+            findUserProfile(activity, authCallback);
+        }
+    }
+
+    public String getAccessToken() {
+        return sharedPreferences.getString(ACS_ACCESS_TOKEN, "");
```
Comment (iaulakh) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/externals/authentication/AADAuthHandler.java:127]: should not be stored in shared preferences

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/views/activities/SignInActivity.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
+        toggleProgress(false);
+        startActivity(intent);
+    }
+
+    private void cacheProfile(final UserProfile profile) {
+        editor.putBoolean(IS_LOGGED_IN, true);
+        editor.putString(USERNAME, profile.getDisplayName());
+        editor.putString(GIVEN_NAME, profile.getDisplayName());
+        editor.putString(ACS_DISPLAY_NAME, profile.getDisplayName());
```
Comment (iaulakh) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/views/activities/SignInActivity.java:99]: Every time user will open app, the request will be make to make sure, user is signed in. I do not think we need to cache this info.

File: AzureCalling/app/src/main/res/layout/activity_start_call.xml [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    <TextView
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:background="@color/fragment_background"
+        android:paddingLeft="16dp"
+        android:textSize="12dp"
+        android:text="Name shown to the others on the call" />
+
+    <LinearLayout
+        android:layout_width="match_parent"
```
Comment (Mrayyan) [inline AzureCalling/app/src/main/res/layout/activity_start_call.xml:63]: Why are we using layout_weight of 1 and also a specific width and height? Shouldn't one of the attribute (width or height) be 0dp?

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/views/activities/SignInActivity.java (hunk 9 above)
Comment (Mohtasim) [reply AzureCalling/app/src/main/java/com/azure/samples/communication/calling/views/activities/SignInActivity.java:99]: No. this is only when user logs in. User do not have to log in every time the app is opened.

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/contracts/SampleErrorMessages.java (hunk 3 above)
Comment (Mohtasim) [reply AzureCalling/app/src/main/java/com/azure/samples/communication/calling/contracts/SampleErrorMessages.java:15]: Noted.

File: AzureCalling/app/src/main/res/layout/activity_start_call.xml (hunk 10 above)
Comment (Mohtasim) [reply AzureCalling/app/src/main/res/layout/activity_start_call.xml:63]: 0dp makes the layout disappears in the smaller screen. wrap content made it visible.

**Review by lpx1989 — DISMISSED** [review]

**Review by Mrayyan — APPROVED** [review]
Nice work 👍

File: AzureCalling/app/src/main/java/com/azure/samples/communication/calling/externals/authentication/AADAuthHandler.java [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+import java.util.function.Consumer;
+
+public class AADAuthHandler {
+
+    private static final String LOG_TAG = AADAuthHandler.class.getSimpleName();
+
+    private final AppSettings appSettings;
+    private ISingleAccountPublicClientApplication mSingleAccountApp;
+    private String accessToken = null;
+    private String[] mScopes = { "User.Read" };
```
Comment (iaulakh) [inline AzureCalling/app/src/main/java/com/azure/samples/communication/calling/externals/authentication/AADAuthHandler.java:46]: If this is required for graph API then we can create new scope variable. As this scope is for AAD.

We can rename this class to AAD+Graph or can create separate classes for handling graph and AAD. Creating separate will be better. Please sync with iOS team.

**Review by ShaunaSong — APPROVED** [review]

**Review by phileo99 — APPROVED** [review]
👍

**Review by iaulakh — APPROVED** [review]

**Pull request merged**